File: apps/frontend/src/voice/realtime-client.ts
/**
 * OpenAI Realtime API client for voice input/output.
 *
 * Handles:
 * - WebSocket connection to Realtime API
 * - Audio streaming (PCM16 24kHz)
 * - Server VAD for automatic speech detection
 * - Transcript extraction (STT)
 * - TTS audio playback via response.audio events
 * - Session configuration with system prompt
 */

const REALTIME_URL = "wss://api.openai.com/v1/realtime";
const MODEL = "gpt-4o-realtime-preview";

export interface RealtimeConfig {
  apiKey: string;
  /** System prompt for context injection */
  systemPrompt?: string;
  /** Voice for TTS output */
  voice?: "alloy" | "echo" | "shimmer" | "ash" | "ballad" | "coral" | "sage" | "verse";
}

export interface RealtimeEvents {
  /** Called when a complete transcript is available */
  onTranscript?: (text: string) => void;
  /** Called with audio data chunks for playback (base64 PCM16 24kHz) */
  onAudio?: (audioBase64: string) => void;
  /** Called when the model finishes speaking */
  onAudioDone?: () => void;
  /** Called when VAD detects speech start */
  onSpeechStart?: () => void;
  /** Called when VAD detects speech end */
  onSpeechEnd?: () => void;
  /** Connection state */
  onConnected?: () => void;
  onDisconnected?: () => void;
  onError?: (error: string) => void;
  /** Called with the refined prompt text from the model */
  onRefinedPrompt?: (prompt: string) => void;
}

export class RealtimeClient {
  private ws: WebSocket | null = null;
  private config: RealtimeConfig;
  private events: RealtimeEvents;
  private disposed = false;

  constructor(config: RealtimeConfig, events: RealtimeEvents = {}) {
    this.config = config;
    this.events = events;
  }

  connect(): void {
    if (this.disposed) return;

    const url = `${REALTIME_URL}?model=${MODEL}`;
    const ws = new WebSocket(url, [
      "realtime",
      `openai-insecure-api-key.${this.config.apiKey}`,
    ]);

    this.ws = ws;

    ws.onopen = () => {
      this.events.onConnected?.();
      this.configureSession();
    };

    ws.onmessage = (event) => {
      try {
        const msg = JSON.parse(event.data as string);
        this.handleMessage(msg);
      } catch {
        // ignore
      }
    };

    ws.onclose = () => {
      this.events.onDisconnected?.();
    };

    ws.onerror = () => {
      this.events.onError?.("Realtime API connection error");
    };
  }

  private configureSession(): void {
    this.send({
      type: "session.update",
      session: {
        modalities: ["text", "audio"],
        instructions: this.config.systemPrompt ?? this.defaultSystemPrompt(),
        voice: this.config.voice ?? "alloy",
        input_audio_format: "pcm16",
        output_audio_format: "pcm16",
        input_audio_transcription: {
          model: "whisper-1",
        },
        turn_detection: {
          type: "server_vad",
          threshold: 0.5,
          prefix_padding_ms: 300,
          silence_duration_ms: 500,
        },
      },
    });
  }

  private defaultSystemPrompt(): string {
    return `You are a voice interface for Teleprompter, a remote Claude Code controller.

Your role:
1. Listen to the user's voice input
2. Clean up and refine it into a clear, actionable prompt for Claude Code
3. Respond briefly to confirm what you understood

Keep responses short and conversational. The refined prompt will be sent to Claude Code automatically.

When the user gives a coding instruction, output it as a clean prompt. For example:
- User says: "um, can you like, fix the bug in the login page, the one where it crashes"
- You respond: "Got it, fixing the login page crash bug."
- Refined prompt: "Fix the bug in the login page that causes a crash"`;
  }

  private handleMessage(msg: any): void {
    switch (msg.type) {
      case "session.created":
      case "session.updated":
        // Session configured successfully
        break;

      case "input_audio_buffer.speech_started":
        this.events.onSpeechStart?.();
        break;

      case "input_audio_buffer.speech_stopped":
        this.events.onSpeechEnd?.();
        break;

      case "conversation.item.input_audio_transcription.completed":
        this.events.onTranscript?.(msg.transcript ?? "");
        break;

      case "response.audio.delta":
        // Streaming audio chunk
        if (msg.delta) {
          this.events.onAudio?.(msg.delta);
        }
        break;

      case "response.audio.done":
        this.events.onAudioDone?.();
        break;

      case "response.text.done":
        // Model's text response (the refined prompt)
        if (msg.text) {
          this.events.onRefinedPrompt?.(msg.text);
        }
        break;

      case "response.done":
        // Extract text from the response output
        if (msg.response?.output) {
          for (const item of msg.response.output) {
            if (item.type === "message") {
              for (const content of item.content ?? []) {
                if (content.type === "text" && content.text) {
                  this.events.onRefinedPrompt?.(content.text);
                }
              }
            }
          }
        }
        break;

      case "error":
        this.events.onError?.(
          msg.error?.message ?? "Realtime API error",
        );
        break;
    }
  }

  /**
   * Send audio data to the Realtime API.
   * @param audioBase64 - Base64-encoded PCM16 24kHz audio
   */
  sendAudio(audioBase64: string): void {
    this.send({
      type: "input_audio_buffer.append",
      audio: audioBase64,
    });
  }

  /**
   * Commit the current audio buffer (trigger processing).
   * Usually not needed with server VAD, but useful for manual mode.
   */
  commitAudio(): void {
    this.send({ type: "input_audio_buffer.commit" });
  }

  /**
   * Send a text message (for text-based interaction with the model).
   */
  sendText(text: string): void {
    this.send({
      type: "conversation.item.create",
      item: {
        type: "message",
        role: "user",
        content: [{ type: "input_text", text }],
      },
    });
    this.send({ type: "response.create" });
  }

  /**
   * Update the system prompt (for context injection).
   */
  updateSystemPrompt(prompt: string): void {
    this.send({
      type: "session.update",
      session: {
        instructions: prompt,
      },
    });
  }

  /**
   * Cancel any in-progress response.
   */
  cancelResponse(): void {
    this.send({ type: "response.cancel" });
  }

  private send(msg: unknown): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(msg));
    }
  }

  disconnect(): void {
    if (this.ws) {
      this.ws.onopen = null;
      this.ws.onmessage = null;
      this.ws.onclose = null;
      this.ws.onerror = null;
      if (
        this.ws.readyState === WebSocket.OPEN ||
        this.ws.readyState === WebSocket.CONNECTING
      ) {
        this.ws.close();
      }
      this.ws = null;
    }
  }

  dispose(): void {
    this.disposed = true;
    this.disconnect();
  }
}

File: ios/Sources/Voice/RealtimeClient.swift
import Foundation
import os

// MARK: - OpenAI Realtime API client
//
// Ported from the Expo app's realtime-client.ts. Connects via WebSocket to
// wss://api.openai.com/v1/realtime with the API key passed in a subprotocol
// header (openai-insecure-api-key.<KEY>), matching the TypeScript original.

private let log = Logger(subsystem: "dev.tpmt.teleprompter", category: "voice.realtime")

private let realtimeURL = "wss://api.openai.com/v1/realtime"
private let realtimeModel = "gpt-4o-realtime-preview"

// MARK: - Event callbacks

struct RealtimeEvents {
    var onConnected: (() -> Void)?
    var onDisconnected: (() -> Void)?
    var onError: ((String) -> Void)?
    var onTranscript: ((String) -> Void)?
    var onAudio: ((String) -> Void)?       // base64 PCM16 24kHz chunk
    var onAudioDone: (() -> Void)?
    var onSpeechStart: (() -> Void)?
    var onSpeechEnd: (() -> Void)?
    var onRefinedPrompt: ((String) -> Void)?  // response.text.done only
}

// MARK: - RealtimeClient

/// WebSocket client for the OpenAI Realtime API.
///
/// Thread-safety: all external calls and callbacks run on the main actor.
/// URLSession delegate callbacks are dispatched to `DispatchQueue.main`.
@MainActor
final class RealtimeClient: NSObject {

    private var task: URLSessionWebSocketTask?
    private var session: URLSession?
    private var events: RealtimeEvents
    private var systemPrompt: String
    private var disposed = false

    init(systemPrompt: String = "", events: RealtimeEvents = RealtimeEvents()) {
        self.systemPrompt = systemPrompt
        self.events = events
    }

    // MARK: - Connection

    func connect(apiKey: String) {
        guard !disposed else { return }

        guard let url = URL(string: "\(realtimeURL)?model=\(realtimeModel)") else {
            events.onError?("Invalid Realtime URL")
            return
        }

        // Subprotocol carries the API key — matches the TypeScript WebSocket constructor:
        //   new WebSocket(url, ["realtime", "openai-insecure-api-key.<KEY>"])
        // URLSession encodes these as the Sec-WebSocket-Protocol header.
        // IMPORTANT: the key is never logged.
        let protocols = ["realtime", "openai-insecure-api-key.\(apiKey)"]

        let cfg = URLSessionConfiguration.default
        let sess = URLSession(configuration: cfg, delegate: self, delegateQueue: .main)
        session = sess

        // webSocketTask(with:protocols:) accepts a URL + protocols list.
        let wsTask = sess.webSocketTask(with: url, protocols: protocols)
        task = wsTask
        wsTask.resume()
        receive()
    }

    // MARK: - Send helpers

    func sendAudio(_ base64: String) {
        send(["type": "input_audio_buffer.append", "audio": base64])
    }

    func commitAudio() {
        send(["type": "input_audio_buffer.commit"])
    }

    func sendText(_ text: String) {
        send([
            "type": "conversation.item.create",
            "item": [
                "type": "message",
                "role": "user",
                "content": [["type": "input_text", "text": text]]
            ]
        ])
        send(["type": "response.create"])
    }

    func updateSystemPrompt(_ prompt: String) {
        systemPrompt = prompt
        send(["type": "session.update", "session": ["instructions": prompt]])
    }

    func cancelResponse() {
        send(["type": "response.cancel"])
    }

    // MARK: - Disconnect / dispose

    func disconnect() {
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        session?.invalidateAndCancel()
        session = nil
    }

    func dispose() {
        disposed = true
        disconnect()
    }

    // MARK: - Private helpers

    private func configureSession() {
        let prompt = systemPrompt.isEmpty ? defaultSystemPrompt() : systemPrompt
        let payload: [String: Any] = [
            "type": "session.update",
            "session": [
                "modalities": ["text", "audio"],
                "instructions": prompt,
                "voice": "alloy",
                "input_audio_format": "pcm16",
                "output_audio_format": "pcm16",
                "input_audio_transcription": ["model": "whisper-1"],
                "turn_detection": [
                    "type": "server_vad",
                    "threshold": 0.5,
                    "prefix_padding_ms": 300,
                    "silence_duration_ms": 500
                ]
            ]
        ]
        send(payload)
    }

    private func defaultSystemPrompt() -> String {
        return """
        You are a voice interface for Teleprompter, a remote Claude Code controller.

        Your role:
        1. Listen to the user's voice input
        2. Clean up and refine it into a clear, actionable prompt for Claude Code
        3. Respond briefly with a spoken confirmation

        Rules:
        - Keep spoken responses SHORT (1-2 sentences max)
        - Output the refined prompt as text in your response
        - The refined prompt will be automatically sent to Claude Code
        - If the user's intent is unclear, ask a brief clarifying question

        Example:
        - User: "um, can you like fix the bug in the login page, the one where it crashes"
        - Your response: "Fixing the login crash bug."
        - (Refined prompt sent to Claude: "Fix the bug in the login page that causes a crash")
        """
    }

    private func send(_ payload: [String: Any]) {
        guard let task, task.state == .running else { return }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let str = String(data: data, encoding: .utf8) else { return }
        task.send(.string(str)) { _ in }
    }

    private func receive() {
        guard let task else { return }
        task.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                switch message {
                case .string(let str):
                    self.handleMessage(str)
                case .data(let data):
                    if let str = String(data: data, encoding: .utf8) {
                        self.handleMessage(str)
                    }
                @unknown default:
                    break
                }
                self.receive()
            case .failure:
                // Socket closed or error — onDisconnected fires via delegate.
                break
            }
        }
    }

    private func handleMessage(_ json: String) {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type_ = obj["type"] as? String else { return }

        switch type_ {
        case "session.created", "session.updated":
            break

        case "input_audio_buffer.speech_started":
            events.onSpeechStart?()

        case "input_audio_buffer.speech_stopped":
            events.onSpeechEnd?()

        case "conversation.item.input_audio_transcription.completed":
            let transcript = obj["transcript"] as? String ?? ""
            events.onTranscript?(transcript)

        case "response.audio.delta":
            if let delta = obj["delta"] as? String, !delta.isEmpty {
                events.onAudio?(delta)
            }

        case "response.audio.done":
            events.onAudioDone?()

        case "response.text.done":
            // Authoritative refined-prompt source (matches TypeScript: only here,
            // NOT from response.done, to avoid duplicates).
            if let text = obj["text"] as? String, !text.isEmpty {
                events.onRefinedPrompt?(text)
            }

        case "response.done":
            // Completion signal — no additional action (refined prompt already emitted
            // via response.text.done above).
            break

        case "error":
            let errObj = obj["error"] as? [String: Any]
            let msg = errObj?["message"] as? String ?? "Realtime API error"
            events.onError?(msg)

        default:
            break
        }
    }
}

// MARK: - URLSessionWebSocketDelegate

extension RealtimeClient: URLSessionWebSocketDelegate {
    nonisolated func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol_: String?
    ) {
        DispatchQueue.main.async { [weak self] in
            guard let self, !self.disposed else { return }
            self.events.onConnected?()
            self.configureSession()
        }
    }

    nonisolated func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        DispatchQueue.main.async { [weak self] in
            self?.events.onDisconnected?()
        }
    }

    nonisolated func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didCompleteWithError error: Error?
    ) {
        guard let error else { return }
        // L12: Route transport errors (e.g. 401 Bad API key — WebSocket upgrade
        // rejected by server) to onError so VoiceStore.lastError is set and
        // the red badge appears. Previously any non-nil error landed in
        // onDisconnected, leaving lastError nil and the failure silent.
        DispatchQueue.main.async { [weak self] in
            self?.events.onError?(error.localizedDescription)
        }
    }
}
